jwst/ramp_fitting/tests/test_ramp_fit.py

```diff
@@ -681,13 +682,39 @@ def test_miri_first_last():
     # Put CR in 1st (0-based) group
     model1.groupdq[0, 1, 1, 1] = JUMP_DET
 
-    new_mod, int_model, opt_model, gls_opt_model = ramp_fit(model1, 1024 * 30000.,
-                                                            True, rnModel, gain, 'OLS',
-                                                            'optimal', 'none')
+    new_mod, int_model, opt_model, gls_opt_model = ramp_fit(
+        model1, 1024 * 30000., True, rnModel, gain, 'OLS', 'optimal', 'none')
 
     np.testing.assert_allclose(new_mod.data, 10. / 3., rtol=1E-5)
 
 
+def test_miri_no_good_pixel():
+    ''' 
+    With no good data, MIRI will remove all groups where all pixels are bad.
+    If all groups are bad, NoneType is returned for all return values from
+    ramp_fit.
+    '''
```

Comment:
How does "no good data" differ from "all groups are bad" ?

Reply:
I don't understand this question.

Reply:
The purpose of the test is to cover the portion of the code where `discard_miri_groups` ends up discarding all groups, i.e., discarding all data.  Because no test covered this code, the bug you found earlier never got detected.  Once I realized this case wasn't being covered I added this test.  I changed the wording of the comments to clarify this better.  Let me know if I need to further clarify the comments.

Additionally, I moved the `discard_miri_groups` to be one of the first functions called.  The discarding of groups in MIRI datasets applies to the dataset as a whole.  For single threading, where it was originally was fine, but for multithreading the effect was to call `discard_miri_groups` for each slice, which is inappropriate.